wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java

```diff
@@ -0,0 +1,161 @@
+          add("notifyNodeListChanged");
+        }
+      });
+
+  private final PluginService pluginService;
+  private final Properties props;
+  private final RdsUtils rdsHelper;
+  private String retrieveInstanceQuery;
+  private String instanceNameCol;
+  private String clusterInstanceTemplate;
+
+  AuroraConnectionDestroyerPlugin(PluginService pluginService, Properties props) {
+    this.pluginService = pluginService;
+    this.props = props;
+    this.rdsHelper = new RdsUtils();
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(String driverProtocol, HostSpec hostSpec, Properties props,
+      boolean isInitialConnection, JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    if (driverProtocol.contains(PG_DRIVER_PROTOCOL)) {
+      this.retrieveInstanceQuery = PG_GET_INSTANCE_NAME_SQL;
+      this.instanceNameCol = PG_INSTANCE_NAME_COL;
+    } else {
+      this.retrieveInstanceQuery = MYSQL_GET_INSTANCE_NAME_SQL;
+      this.instanceNameCol = MYSQL_GET_INSTANCE_NAME_COL;
+    }
+
+    final Connection conn = connectFunc.call();
+
+    if (conn != null) {
+      final HostSpec currentHostSpec = this.pluginService.getCurrentHostSpec();
+      if (rdsHelper.isRdsClusterDns(currentHostSpec.getHost())) {
+        currentHostSpec.addAlias(getInstanceEndpoint(conn, currentHostSpec));
+      }
+    }
+
+    OpenedConnectionTracker.populateOpenedConnectionQueue(hostSpec, conn);
+    OpenedConnectionTracker.logOpenedConnections();
+
+    return conn;
+  }
+
+  private String getInstanceEndpointPattern(final String url) {
+    if (StringUtils.isNullOrEmpty(this.clusterInstanceTemplate)) {
+      this.clusterInstanceTemplate = AuroraHostListProvider.CLUSTER_INSTANCE_HOST_PATTERN.getString(this.props) == null
+          ? rdsHelper.getRdsInstanceHostPattern(url)
+          : AuroraHostListProvider.CLUSTER_INSTANCE_HOST_PATTERN.getString(this.props);
+    }
+
+    return this.clusterInstanceTemplate;
+  }
+
+  @Override
+  public <T, E extends Exception> T execute(Class<T> resultClass, Class<E> exceptionClass,
+      Object methodInvokeOn, String methodName, JdbcCallable<T, E> jdbcMethodFunc,
+      Object[] jdbcMethodArgs) throws E {
+    final HostSpec originalHost = this.pluginService.getCurrentHostSpec();
+    try {
+      T result = jdbcMethodFunc.call();
+      if ((methodName.contains(METHOD_CLOSE) || methodName.contains(METHOD_ABORT))) {
+        OpenedConnectionTracker.invalidateCurrentConnection(originalHost, this.pluginService.getCurrentConnection());
+      }
+      return result;
+    } catch (Exception e) {
+      OpenedConnectionTracker.logOpenedConnections();
+      if (e instanceof FailoverSQLException) {
+        OpenedConnectionTracker.invalidateAllConnections(originalHost);
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void notifyNodeListChanged(Map<String, EnumSet<NodeChangeOptions>> changes) {
```

Comment:
During failover this notification may be quite noisy and gives false positive. Can we reply just on failover exceptions instead?